compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt

```diff
@@ -909,7 +909,7 @@ class MethodInliner(
         }
 
         if (inliningContext.isInliningLambda && inliningContext.lambdaInfo is IrExpressionLambda) {
-            val capturedVars = inliningContext.lambdaInfo.capturedVars
+            val capturedVars = parameters.captured.map { it.desc }
```

Comment:
@pyos My apologies for delay in review, actually I can't catch root cause in this change. But after fresh view there are some results:

I'm observing next exception:
```
Caused by: org.jetbrains.kotlin.codegen.inline.InlineException: Transformer for _2Kt$complicatedCast$1$1: couldn't inline method call
Cause: f (Ljava/lang/Object;)Ljava/lang/Object;:
    @Lorg/jetbrains/annotations/Nullable;() // invisible, parameter 0
   L0
    LINENUMBER 8 L0
    ALOAD 1
    ASTORE 2
   L1
    LINENUMBER 9 L1
    GETSTATIC _2Kt$complicatedCast$1.$$$$this : Ljava/lang/Object;
    ICONST_1
    LDC "A"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.reifiedOperationMarker (ILjava/lang/String;)V
    CHECKCAST java/lang/Object
    ARETURN
```
And there is very strange     `_2Kt$complicatedCast$1.$$$$this : Ljava/lang/Object;` access inside 'f' function. That shouldn't be at all. This code should be executed only for root inline lambdas, `isInliningLambda` is inherited by inner context of anonymoys object transformation (yes, bad naming). Switching condition (using  original code for this line) makes test pass but add some new failings)

Reply:
>Switching condition (using original code for this line) makes test pass but add some new failings

What condition did you try? Simply adding ` && !inliningContext.parent!!.isInliningLambda` to filter out contexts inside the lambda does seem to work.

In general, like I said in the PR description, the object does not actually need to be regenerated, since it's an IR-generated object inside a lambda and thus KT-28064 applies. The only reason this code is reachable with a context nested inside a lambda at all is that `AnonymousObjectTransformationInfo.shouldRegenerate` returns true if the object uses a reified type parameter, which is does here, even though the type parameter is defined outside the root inlining context and thus does not need to be reified yet. (The "needClassReification" marker unfortunately does not carry any information about the type parameters themselves, so it's impossible to find that out without first scanning the entire class.) Thus, literally any change that prevents the code from modifying anonymous objects inside lambdas (like the PR itself, or changing the condition to not do anything in nested contexts, or even having `MethodInliner.isAlreadyRegenerated` return true if `inliningContext.lambdaInfo is IrExpressionLambda`) will fix the issue.

Reply:
Yes, I've tried mentioned condition. I will prefer solution with patched condition in 'transformCaptured' cause it will allow to remove dangerous code (that could affect another regenerations triggered by any (right or wrong) reasons)

Reply:
Done, though the regeneration condition should still be eventually modified to comply with KT-28064

Reply:
Yes, indeed. It's best to create separate pr for this